react-components/src/hooks/useSearchMappedEquipmentFDM.tsx

```diff
@@ -308,30 +421,103 @@ function createInModelsFilter(models: AddModelOptions[]): { in: any } {
+    queryResult.items.mapped_edges as EdgeItem[],
+    models,
+    spacesToSearch
+  );
+
   for (const searchResult of searchResults) {
     if (searchResult.instances.length === 0) {
       continue;
     }
 
-    const isMappedFilter = createIsMappedFilter(searchResult.instances, models);
-    const mappedEdges = await fdmSdk.filterAllInstances(
-      isMappedFilter,
-      'edge',
-      SYSTEM_3D_EDGE_SOURCE
+    const filteredInstances = searchResult.instances.filter((instance) =>
+      checkInstanceWithMappedEquipmentMaps(
+        mappedEquipmentFirstLevelMap,
+        equipmentSecondLevelMap,
+        instance
+      )
     );
 
-    const mappedEquipmentMap = createMappedEquipmentMap(mappedEdges.instances);
-    const filteredInstances = searchResult.instances.filter((instance) => {
-      const key = `${instance.space}/${instance.externalId}`;
-      return mappedEquipmentMap[key] !== undefined;
-    });
-
     filteredSearchResults.push({ view: searchResult.view, instances: filteredInstances });
   }
 
   return filteredSearchResults;
 }
+
+function checkInstanceWithMappedEquipmentMaps(
+  mappedEquipmentFirstLevelMap: Record<string, EdgeItem>,
+  equipmentSecondLevelMap: Record<string, EdgeItem>,
+  instance: NodeItem
+): boolean {
+  const key = `${instance.space}/${instance.externalId}`;
+  const directRelationProperties = getDirectRelationProperties(instance);
+  const isMappedFirstLevel = mappedEquipmentFirstLevelMap[key] !== undefined;
+  const isSecondLevelWithEdge = equipmentSecondLevelMap[key] !== undefined;
+  const isSecondLevelWithDirectRelation = directRelationProperties.length > 0;
+
+  if (isMappedFirstLevel) {
+    return true;
+  }
+
+  if (isSecondLevelWithEdge) {
+    const { space, externalId } = equipmentSecondLevelMap[key].endNode;
+
+    const secondLevelKey = `${space}/${externalId}`;
+    const isMappedWithEdge = mappedEquipmentFirstLevelMap[secondLevelKey] !== undefined;
+
+    return isMappedWithEdge;
+  }
+
+  if (isSecondLevelWithDirectRelation) {
+    const isMappedWithDirectRelation = directRelationProperties.some(
+      ({ externalId, space }) =>
+        mappedEquipmentFirstLevelMap[`${space}/${externalId}`] !== undefined
+    );
+
+    return isMappedWithDirectRelation;
+  }
+
+  return false;
+}
+
+function getRevisionIdFromEdge(edge: EdgeItem): string | undefined {
+  return (edge.properties as any)?.[SYSTEM_SPACE_3D_SCHEMA]?.[
+    `${SYSTEM_3D_EDGE_SOURCE.externalId}/${SYSTEM_3D_EDGE_SOURCE.version}`
+  ]?.revisionId?.toString();
+}
+
+function getDirectRelationProperties(searchResultNode: NodeItem): DmsUniqueIdentifier[] {
+  const directRelations: DmsUniqueIdentifier[] = [];
+  const nodeProperties = searchResultNode.properties;
+
+  Object.keys(nodeProperties).forEach((propertyKey) => {
+    const { space, externalId } = nodeProperties[propertyKey] as any;
+
+    if (space !== undefined && externalId !== undefined) {
+      directRelations.push({ space, externalId });
+    }
```

Comment:
Yeah, but it will just check them against the map that was created from the query, so even if it won't be a direct relation it won't break, just add a tiny overhead for checks.